### DISCUSSION TITLE
opentelemetry, how to disable some of the default spans?

### DISCUSSION THREAD

**Discussion by ceastman-r7** [vercel/next.js#58614] [Help]
### Summary

Specifically I would like to disable a couple of apis from having spans created for them. namely the /liveness and /readiness endpoints that kubernetes uses to see if the app is ready and alive. these endpoints get hit alot from the kubernetes api server.

I followed this doc
https://nextjs.org/docs/app/building-your-application/optimizing/open-telemetry#manual-opentelemetry-configuration

**Comment by tomdohnal** [comment]
I haven't found a way to disable producing the spans but I've been experimenting with using a custom `Sampler` to filter out Next.js spans:

```typescript

class CustomSampler implements Sampler {
  shouldSample(
    context: Context,
    traceId: string,
    spanName: string,
    spanKind: SpanKind,
    attributes: Attributes,
  ): SamplingResult {
    if (attributes['next.span_type']) {
      return {
        decision: SamplingDecision.NOT_RECORD,
      }
    }

    return {
      decision: SamplingDecision.RECORD_AND_SAMPLED,
    }
  }
}
```

I think you could adopt this for your usecase by looking into the `attributes` (probably something like `attributes['next.route'] !== '/readiness'`).

Although it'd be defintely appreciated to have a way to control span generation baked into Next.js itself.

**Comment by stefann42** [comment]
Another workaround is to filter at the level of the exporter. For example, you can override `OTLPTraceExporter.send` in your own derived class: 

```
class OTLPTraceExporterWithFilter extends OTLPTraceExporter {
  override send(objects: ReadableSpan[], onSuccess: () => void, onError: (error: OTLPExporterError) => void): void {
    const objectsToSend = objects.filter(i => !Object.keys(i.attributes).some(key => key.startsWith("next.")))
    super.send(objectsToSend, onSuccess, onError);
  }
}
```

Above is a very coarse criteria for dropping everything that comes from the next.js framework itself, but you can tweak the filter criteria to whatever you need. 

As of next.js 15.0.1 default spans are documented at Default Spans https://nextjs.org/docs/pages/building-your-application/optimizing/open-telemetry#default-spans-in-nextjs and the resulting `Span` object that comes into the exporter looks like this:

```
Span {
    attributes: {
      'next.clientComponentLoadCount': 14,
      'next.span_type': 'NextNodeServer.clientComponentLoading'
    },
    links: [],
    events: [],
    _droppedAttributesCount: 0,
    _droppedEventsCount: 0,
    _droppedLinksCount: 0,
    status: { code: 0 },
    endTime: [ 1730167368, 619089053 ],
    _ended: true,
    _duration: [ 0, 431007800 ],
    name: 'NextNodeServer.clientComponentLoading',
    _spanContext: {
      traceId: 'aea97a60b42591fdc822dfa50f4768e6',
      spanId: '7a910d1be40c6eb7',
      traceFlags: 1,
      traceState: undefined
    },
    parentSpanId: undefined,
    kind: 0,
    _performanceStartTime: 16518.2834,
    _performanceOffset: -0.578369140625,
    _startTimeProvided: true,
    startTime: [ 1730167368, 188081253 ],
    resource: Resource {
      _attributes: [Object],
      asyncAttributesPending: false,
      _syncAttributes: [Object],
      _asyncAttributesPromise: [Promise]
    },
    instrumentationLibrary: { name: 'next.js', version: '0.0.1', schemaUrl: undefined },
    _spanLimits: {
      attributeValueLengthLimit: Infinity,
      attributeCountLimit: 128,
      linkCountLimit: 128,
      eventCountLimit: 128,
      attributePerEventCountLimit: 128,
      attributePerLinkCountLimit: 128
    },
    _attributeValueLengthLimit: Infinity,
    _spanProcessor: MultiSpanProcessor { _spanProcessors: [Array] }
  }
```

**Reply by trieu-hphuc96** [reply]
It's working for me, thanks for the solution ❤️

**Comment by Marlrus** [comment]
@stefann42 Thanks for the code!

I have an `api/health` endpoint that gets called 3 times every 10s and was creating traces. I used an attribute to target the Spans to remove but that removed some of the Spans as they hand no single unifying Span Attribute under the Node Auto Instrumentation.

I modified the code to remove all related spans when any span has an attribute which I chose to make `exclude: true`. When I find a span, I get the `traceId` from it and remove all spans with the same `traceId` which effectively removes all related spans.

``` typescript
class OTLPTraceExporterWithFilter extends OtelOtlpTraceExporter.OTLPTraceExporter {
  override send(
    spans: OtelTraceBaseSdk.ReadableSpan[],
    onSuccess: () => void,
    onError: (error: OtelOtlpExporterBase.OTLPExporterError) => void,
  ): void {
    const spansToExclude = spans.filter((span) => span.attributes.exclude === 'true');
    const traceIdsToExclude = spansToExclude.map((span) => span.spanContext().traceId);

    const spanToSend = spans.filter(
      (span) => !traceIdsToExclude.includes(span.spanContext().traceId),
    );

    super.send(spanToSend, onSuccess, onError);
  }
}
```

With this, if I want to exclude all Spans related to a request/flow, I just append the attribute `exclude: 'true'` to it and its out.

I have a custom logger that integrates with OTEL but you should be able to get the Span from the active context and append the attributes directly:

```typescript
const activeContext = otel.context.active();

const activeSpan = otel.trace.getSpan(activeContext);

activeSpan.setAttributes({ exclude: 'true' })
```
